dotnet/src/Connectors/Connectors.Memory.Pinecone/PineconeVectorStoreRecordMapper.cs

```diff
@@ -0,0 +1,193 @@
+            if (propertyValue != null)
+            {
+                metadata[propertyName] = ConvertToMetadataValue(propertyValue);
+            }
+        }
+
+        var valuesObject = this._vectorPropertyInfo.GetValue(dataModel);
+        if (valuesObject is not ReadOnlyMemory<float> values)
+        {
+            throw new VectorStoreRecordMappingException($"Vector property {this._vectorPropertyInfo.Name} on provided record of type {typeof(TRecord).FullName} may not be null.");
+        }
+
+        // TODO: what about sparse values?
+        var result = new Vector
+        {
+            Id = (string)keyObject,
+            Values = values.ToArray(),
+            Metadata = metadata,
+            SparseValues = null
+        };
+
+        return result;
+    }
+
+    /// <inheritdoc />
+    public TRecord MapFromStorageToDataModel(Vector storageModel, StorageToDataModelMapperOptions options)
+    {
+        var keyPropertyName = this._storagePropertyNames[this._keyPropertyInfo.Name];
+
+        var outputJsonObject = new JsonObject
+        {
+            { this._keyPropertyInfo.Name, JsonValue.Create(storageModel.Id) },
+        };
+
+        if (options?.IncludeVectors is true)
+        {
+            var propertyName = this._storagePropertyNames[this._vectorPropertyInfo.Name];
+            outputJsonObject.Add(this._vectorPropertyInfo.Name, new JsonArray(storageModel.Values.Select(x => JsonValue.Create(x)).ToArray()));
+        }
+
+        if (storageModel.Metadata != null)
+        {
+            foreach (var payloadProperty in this._payloadPropertiesInfo)
+            {
+                var propertyName = this._storagePropertyNames[payloadProperty.Name];
+                if (storageModel.Metadata.TryGetValue(propertyName, out var value))
+                {
+                    outputJsonObject.Add(payloadProperty.Name, ConvertFromMetadataValueToJsonNode(value));
+                }
+            }
+        }
+
+        return outputJsonObject.Deserialize<TRecord>()!;
+    }
+
+    private static JsonNode? ConvertFromMetadataValueToJsonNode(MetadataValue metadataValue)
+        => metadataValue.Inner switch
+        {
+            null => null,
+            bool boolValue => JsonValue.Create(boolValue),
+            string stringValue => JsonValue.Create(stringValue),
+            int intValue => JsonValue.Create(intValue),
+            long longValue => JsonValue.Create(longValue),
+            float floatValue => JsonValue.Create(floatValue),
+            double doubleValue => JsonValue.Create(doubleValue),
+            decimal decimalValue => JsonValue.Create(decimalValue),
+            MetadataValue[] array => new JsonArray(array.Select(ConvertFromMetadataValueToJsonNode).ToArray()),
+            _ => throw new VectorStoreRecordMappingException($"Unsupported metadata type: '{metadataValue.Inner?.GetType().FullName}'."),
+        };
+
+    private static MetadataValue ConvertToMetadataValue(object? sourceValue)
```

Comment:
This is not the first time the users face with a UX issue around creating `MetadataValue`s.

I have added [`MetadataValue.TryCreate`](https://github.com/neon-sunset/Pinecone.NET/blob/aa68e65f3b46c195a0922221156d549da0b27622/src/Types/VectorTypes.cs#L124) to address this.

Reply:
will take advantage of this once we move to the new version of the SDK